### PR TITLE
stm32 qspi driver supports dual quad nor in Memorymapped mode

### LINKED ISSUE #73025: stm32H7  cannot XiP  from dual quad nor in Memorymapped mode
When executing the samples/application_development/code_relocation_nocopy/ on the stm32h745i_disco//m7 disco kit,
the function_in_ext_flash cannot execute correctly fro theexternal NOR flash

The external NOR flash is configured in memoryMapped mode with 
CONFIG_STM32_MEMMAP=y

The samples/application_development/code_relocation_nocopy/ is build with quadspi memory region set at 0x90000000
The function_in_ext_flash is downloaded  in ext flash, but execution fails 

```
*** Booting Zephyr OS build v3.6.0-4147-gadeb19b30cbf ***                                     
Address of main function 0x800066d                                                            
[00:00:00.000,000] <err> os: ***** BUS FAULT *****                                            
[00:00:00.000,000] <err> os:   Instruction bus error                                          
[00:00:00.000,000] <err> os: r0/a1:  0x00000003  r1/a2:  0x00000000  r2/a3:  0x00000004       
[00:00:00.000,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x08000681       
[00:00:00.000,000] <err> os:  xpsr:  0x61000000                                               
[00:00:00.000,000] <err> os: Faulting instruction address (r15/pc): 0x90000000                
[00:00:00.000,000] <err> os: >>> ZEPHYR FATAL ERROR 27: Unknown error on CPU 0                
[00:00:00.000,000] <err> os: Current thread: 0x24000800 (unknown)                             
[00:00:00.042,000] <err> os: Halting system     

```

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows) ubuntu
 - Toolchain (e.g Zephyr SDK, ...) 0.16.3
 - Commit SHA or Version used  adeb19b30cbfb3552cb2d19c6c6676a872afe6c0

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is adding the support of the DFM (DualFlash Mode) to the stm32QSPI driver.
This mode activates two external quad-NOR memories simultaneously to double throughput and capacity
Especially on stm32h745i and stm32h747i disco kits where two MT25QL512 quad-NOR are mounted on 2 separated qspi IO banks.
Introducing a new property <dual-flash> in the DTS of the "st,stm32-qspi" node.
Check the <dual-flash>  boolean property’s truth value (with DT_PROP) to set the DFM.

Requires https://github.com/zephyrproject-rtos/zephyr/pull/68377  to flash with STM32CubeProgrammer

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/73025

### CODE REVIEW

**Comment by FRASTM** [comment]
`west build -p always -b stm32h745i_disco//m7 samples/application_development/code_relocation_nocopy/`
west flash hex file on internal and external memories
```
*** Booting Zephyr OS build v3.6.0-4153-gb1647f830c0b ***                               
Address of main function 0x800066d                                                      
Address of function_in_ext_flash 0x90000001                                             
Address of var_ext_sram_data 0x240000e0 (10)                                            
Address of function_in_sram 0x24000001                                                  
Address of var_sram_data 0x240000e4 (10)                                                
Hello World! stm32h745i_disco  
```

**Comment by toonst** [comment]
Tried this on the `stm32h750b_dk`, got a `Unaligned memory access` fault:

```
*** Booting Zephyr OS build v3.6.0-4151-ga26e85ed51da ***
Address of main function 0x800066d
[00:00:00.000,000] <err> os: ***** USAGE FAULT *****
[00:00:00.000,000] <err> os:   Unaligned memory access
[00:00:00.000,000] <err> os: r0/a1:  0x00000064  r1/a2:  0xcf7bcf7b  r2/a3:  0xcf7bcf7b
[00:00:00.000,000] <err> os: r3/a4:  0xcf7bcf7b r12/ip:  0x00000000 r14/lr:  0x08000681
[00:00:00.000,000] <err> os:  xpsr:  0x01000000
[00:00:00.000,000] <err> os: Faulting instruction address (r15/pc): 0x0000759a
[00:00:00.000,000] <err> os: >>> ZEPHYR FATAL ERROR 31: Unknown error on CPU 0
[00:00:00.000,000] <err> os: Current thread: 0x24000720 (unknown)
[00:00:00.000,000] <err> os: Halting system
```

**Comment by FRASTM** [comment]
> Tried this on the `stm32h750b_dk`, got a fault:
> 
> ```
> *** Booting Zephyr OS build v3.6.0-4155-ge863717bd6e4 ***
> Address of main function 0x800066d
> [00:00:00.000,000] <err> os: ***** BUS FAULT *****
> [00:00:00.000,000] <err> os:   Instruction bus error
> [00:00:00.000,000] <err> os: r0/a1:  0x080064de  r1/a2:  0x08006516  r2/a3:  0x00007001
> [00:00:00.000,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x0000700d
> [00:00:00.000,000] <err> os:  xpsr:  0x60000000
> [00:00:00.000,000] <err> os: Faulting instruction address (r15/pc): 0x8d56a448
> [00:00:00.000,000] <err> os: >>> ZEPHYR FATAL ERROR 27: Unknown error on CPU 0
> [00:00:00.000,000] <err> os: Current thread: 0x24000720 (unknown)
> [00:00:00.000,000] <err> os: Halting system
> ```
> 
> Tried flashing with `west flash --runner stm32cubeprogrammer --extload MT25TL01G_STM32H750B-DISCO.stldr --tool-opt=0x90000000` using the mentioned PR.

To Pass the samples/application_development/code_relocation_nocopy/ on the stm32h750b_dk disco kit, you need
- CONFIG_STM32_MEMMAP=y in a ./boards/stm32h750b_dk.conf
-  a redefinition of the EXTMEM MPU region with ATTR_MPU_IO attributes

```
*** Booting Zephyr OS build v3.6.0-4153-g81aedfe06d29 ***
Address of main function 0x800066d
Address of function_in_ext_flash 0x90000001
Address of var_ext_sram_data 0x240000e0 (10)
Address of function_in_sram 0x24000001                                                                                                                                       
Address of var_sram_data 0x240000e4 (10)                                                                                                                                     
Hello World! stm32h750b_dk     
```

**Comment by FRASTM** [comment]
adding the support of the stm32h750b_dk (dual quad-flash) for the samples/application_development/code_relocation_nocopy/

**Comment by FRASTM** [comment]
- Rebase on c1400536548bb861c630f81d8aced864f1a85ca1
- Tested with the stm32h745i_disco//m7 or stm32h750b_dk dual-flash
- Tested with the stm32f746g_disco (single quad)
The stm32CubeProgrammer (> v2.16.0) is used as a runner to flash zephyr.hex file

File: dts/bindings/qspi/st,stm32-qspi.yaml [hunk 1]
```diff
@@ -59,12 +59,22 @@ properties:
...
+      When true, the FLash ID number <flash-id> is useless.
+
   flash-id:
     type: int
     description: |
       FLash ID number. This number, if defined, helps to select the right
-      QSPI GPIO banks (defined as 'quadspi_bk[12]' in pinctrl property)
+      QSPI GPIO banks (defined as 'quadspi_bk[1/2]' in pinctrl property)
       to communicate with flash memory.
+      Valid if the <dual-flash> is false.
```
Comment (erwango) [inline dts/bindings/qspi/st,stm32-qspi.yaml:77]:
```
      Valid only if the <dual-flash> is not set.
```

Comment (FRASTM) [reply dts/bindings/qspi/st,stm32-qspi.yaml]: changed

**Comment by FRASTM** [comment]
Fixing conflict when rebasing on a15af0be9ff64436e94f29de48e394c9d1f9f6f8

File: dts/bindings/qspi/st,stm32-qspi.yaml [hunk 2]
```diff
@@ -59,12 +59,22 @@ properties:
...
          dma-names = "tx_rx";
 
+  dual-flash:
+    type: boolean
+    description: |
+      configuration to enable the dual flash mode of the QSPI peripheral
+      where two external quad SPI Flash memories (FLASH 1 and FLASH 2) are used
+      in order to send/receive 8 bits (or 16 bits in DDR mode) every cycle,
+      effectively doubling the throughput as well as the capacity.
+      When true, the FLash ID number <flash-id> is useless.
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply dts/bindings/qspi/st,stm32-qspi.yaml]: > 
corrected

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**